### PR TITLE
MG-2048 - Implement Personal Access Tokens (PATs)

### LINKED ISSUE #2048: Implement Personal Access Tokens (PATs)
API keys will be used for communication between services. This can be done even as a separate service, but we'll first need some investigation and specs before implementing it. @arvindh123 Please write all the findings in the comment section of this issue.

### PR DESCRIPTION
**Pull request by nyagamunene**
# What type of PR is this?
This is a feature because it adds the following functionality: It add personal access token.

## What does this do?
It adds personal access token.

## Which issue(s) does this PR fix/relate to?

- Resolves #2048 

## Have you included tests for your changes?
Yes

## Did you document any new/modified feature?
No

### CODE REVIEW

File: auth/api/grpc/auth/client.go [hunk 1]
```diff
@@ -109,3 +109,4 @@ func encodeAuthorizeRequest(_ context.Context, grpcReq interface{}) (interface{}
 		Object:      req.Object,
 	}, nil
 }
+
```
Comment (arvindh123) [inline auth/api/grpc/auth/client.go:148]: Remove this line

File: auth/api/grpc/auth/requests.go [hunk 2]
```diff
@@ -49,3 +49,19 @@ func (req authReq) validate() error {
...
+	operation                string
+	entityIDs                []string
+}
+
+func (req authPATReq) validate() error {
+	if req.paToken == "" {
+		return apiutil.ErrBearerToken
+	}
+	return nil
+}
```
Comment (arvindh123) [inline auth/api/grpc/auth/requests.go]: This was not used anywhere

File: pkg/authn/authsvc/authn.go [hunk 3]
```diff
@@ -42,5 +42,5 @@ func (a authentication) Authenticate(ctx context.Context, token string) (authn.S
 	if err != nil {
 		return authn.Session{}, errors.Wrap(errors.ErrAuthentication, err)
 	}
-	return authn.Session{DomainUserID: res.GetId(), UserID: res.GetUserId(), DomainID: res.GetDomainId()}, nil
+	return authn.Session{ID: res.GetId(), UserID: res.GetUserId(), DomainID: res.GetDomainId()}, nil
```
Comment (arvindh123) [inline pkg/authn/authsvc/authn.go]: Please revert to DomainUserID,This affect everywhere, since we are using DomainUserID in all places

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: pkg/authn/authsvc/authn.go (hunk 3 above)
Comment (nyagamunene) [reply pkg/authn/authsvc/authn.go]: Since domain was removed from token this will be empty

File: auth/api/grpc/auth/requests.go [hunk 4]
```diff
@@ -49,3 +50,23 @@ func (req authReq) validate() error {
...
+	platformEntityType       auth.PlatformEntityType
+	optionalDomainID         string
+	optionalDomainEntityType auth.DomainEntityType
+	operation                auth.OperationType
+	entityIDs                []string
+}
+
+func (req authPATReq) validate() error {
+	if req.userID == "" {
+		return apiutil.ErrBearerToken
```
Comment (felixgateru) [inline auth/api/grpc/auth/requests.go]: Why this error instead of `ErrMissingID` or `ErrAuthorization`

File: auth/api/grpc/auth/requests.go [hunk 5]
```diff
@@ -49,3 +50,23 @@ func (req authReq) validate() error {
...
+	}
+	if req.patID == "" {
+		return apiutil.ErrBearerToken
```
Comment (felixgateru) [inline auth/api/grpc/auth/requests.go]: same as above.

File: auth/api/grpc/auth/endpoint.go [hunk 6]
```diff
@@ -50,3 +66,18 @@ func authorizeEndpoint(svc auth.Service) endpoint.Endpoint {
...
+		if err := req.validate(); err != nil {
+			return authorizeRes{}, err
+		}
+		err := svc.AuthorizePAT(ctx, req.userID, req.patID, req.platformEntityType, req.optionalDomainID, req.optionalDomainEntityType, req.operation, req.entityIDs...)
+		if err != nil {
+			return authorizeRes{authorized: false}, err
+		}
+		return authorizeRes{authorized: true}, nil
+	}
+}
```
Comment (felixgateru) [inline auth/api/grpc/auth/endpoint.go:83]: We will need to add tests for these changes

File: auth/api/http/pats/responses.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+func (res addPatScopeEntryRes) Empty() bool {
+	return false
+}
+
+type removePatScopeEntryRes struct {
+	auth.Scope
+}
+
+func (res removePatScopeEntryRes) Code() int {
+	return http.StatusAccepted
```
Comment (felixgateru) [inline auth/api/http/pats/responses.go]: Why this status code?I think this is used for requests that have been accepted for processing,but not been completed

File: auth/api/http/pats/responses.go [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+func (res revokePatSecretRes) Empty() bool {
+	return true
+}
+
+type addPatScopeEntryRes struct {
+	auth.Scope
+}
+
+func (res addPatScopeEntryRes) Code() int {
+	return http.StatusAccepted
```
Comment (felixgateru) [inline auth/api/http/pats/responses.go]: Same as below

File: auth/api/http/pats/transport.go [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+	mux.Route("/pats", func(r chi.Router) {
+		r.Post("/", kithttp.NewServer(
+			createPATEndpoint(svc),
+			decodeCreatePATRequest,
+			api.EncodeResponse,
+			opts...,
+		).ServeHTTP)
+
+		r.Get("/{id}", kithttp.NewServer(
+			(retrievePATEndpoint(svc)),
```
Comment (felixgateru) [inline auth/api/http/pats/transport.go]: Are the brackets necessary?

File: auth/api/http/pats/transport.go [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+			decodeRetrievePATRequest,
+			api.EncodeResponse,
+			opts...,
+		).ServeHTTP)
+
+		r.Patch("/{id}/name", kithttp.NewServer(
```
Comment (felixgateru) [inline auth/api/http/pats/transport.go]: You can arrange the routes by grouping

File: auth/api/logging.go [hunk 11]
```diff
@@ -124,3 +124,253 @@ func (lm *loggingMiddleware) Authorize(ctx context.Context, pr policies.Policy)
...
+		args := []any{
+			slog.String("duration", time.Since(begin).String()),
+			slog.String("name", name),
+			slog.String("description", description),
+			slog.String("pat_duration", duration.String()),
+			slog.String("scope", scope.String()),
+		}
+		if err != nil {
+			args = append(args, slog.Any("error", err))
+			lm.logger.Warn("Create PAT failed to complete successfully", args...)
```
Comment (felixgateru) [inline auth/api/logging.go]:
```suggestion
			lm.logger.Warn("Create PAT failed", args...)
``

File: auth/api/logging.go [hunk 12]
```diff
@@ -124,3 +124,253 @@ func (lm *loggingMiddleware) Authorize(ctx context.Context, pr policies.Policy)
...
+func (lm *loggingMiddleware) UpdatePATDescription(ctx context.Context, token, patID, description string) (pa auth.PAT, err error) {
+	defer func(begin time.Time) {
+		args := []any{
+			slog.String("duration", time.Since(begin).String()),
+			slog.String("pat_id", patID),
+			slog.String("description", description),
+		}
+		if err != nil {
+			args = append(args, slog.Any("error", err))
+			lm.logger.Warn("Update PAT description failed to complete successfully", args...)
```
Comment (felixgateru) [inline auth/api/logging.go]: Same as above

File: internal/api/common.go [hunk 13]
```diff
@@ -128,7 +128,8 @@ func EncodeError(_ context.Context, err error, w http.ResponseWriter) {
 	case errors.Contains(err, svcerr.ErrAuthorization),
 		errors.Contains(err, svcerr.ErrDomainAuthorization),
 		errors.Contains(err, bootstrap.ErrExternalKey),
-		errors.Contains(err, bootstrap.ErrExternalKeySecure):
+		errors.Contains(err, bootstrap.ErrExternalKeySecure),
+		errors.Contains(err, apiutil.ErrUnsupportedTokenType):
```
Comment (felixgateru) [inline internal/api/common.go]: Should this not be a 401 instead

**Review by felixgateru — CHANGES_REQUESTED** [review]

**Review by felixgateru — DISMISSED** [review]

File: auth/api/http/keys/transport.go [hunk 14]
```diff
@@ -64,8 +72,13 @@ func decodeIssue(_ context.Context, r *http.Request) (interface{}, error) {
 }
 
 func decodeKeyReq(_ context.Context, r *http.Request) (interface{}, error) {
+	token := apiutil.ExtractBearerToken(r)
+	if strings.HasPrefix(token, patPrefix) {
+		return nil, apiutil.ErrUnsupportedTokenType
+	}
```
Comment (arvindh123) [inline auth/api/http/keys/transport.go]: This check is not required, Because in the above line 
`apiutil.ExtractBearerToken` have check like this `strings.HasPrefix(token, BearerPrefix)`
This will return empty string if token doesn't start with `Bearer`

File: auth/api/http/keys/transport.go [hunk 15]
```diff
@@ -55,7 +58,12 @@ func decodeIssue(_ context.Context, r *http.Request) (interface{}, error) {
 		return nil, apiutil.ErrUnsupportedContentType
 	}
 
-	req := issueKeyReq{token: apiutil.ExtractBearerToken(r)}
+	token := apiutil.ExtractBearerToken(r)
+	if strings.HasPrefix(token, patPrefix) {
+		return nil, apiutil.ErrUnsupportedTokenType
+	}
```
Comment (arvindh123) [inline auth/api/http/keys/transport.go]: This check is not required, Because in the above line 
`apiutil.ExtractBearerToken` have check like this `strings.HasPrefix(token, BearerPrefix)`
This will return empty string if token doesn't start with `Bearer`

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: internal/api/auth.go [hunk 16]
```diff
@@ -24,11 +29,20 @@ func AuthenticateMiddleware(authn smqauthn.Authentication, domainCheck bool) fun
 				EncodeError(r.Context(), apiutil.ErrBearerToken, w)
 				return
 			}
-
-			resp, err := authn.Authenticate(r.Context(), token)
-			if err != nil {
-				EncodeError(r.Context(), err, w)
-				return
+			var resp smqauthn.Session
+			var err error
```
Comment (arvindh123) [inline internal/api/auth.go:31]: @nyagamunene 
Have you tested this? 
I think code will not reach here if we sent `PAT` in Bearer token

**Review by arvindh123 — CHANGES_REQUESTED** [review]

**Review by arvindh123 — CHANGES_REQUESTED** [review]
Please check for PAT exipry in AuthN

File: auth/api/http/keys/transport.go (hunk 15 above)
Comment (arvindh123) [reply auth/api/http/keys/transport.go]: I think, this will be handled by service layer, Please check and remove if not needed

File: internal/api/auth.go (hunk 16 above)
Comment (nyagamunene) [reply internal/api/auth.go:31]: It does reach to this code. I have used PAT in authorization and I send the PAT in bearer token

File: auth/api/http/keys/transport.go [hunk 17]
```diff
@@ -18,7 +18,10 @@ import (
 	kithttp "github.com/go-kit/kit/transport/http"
 )
 
-const contentType = "application/json"
+const (
+	contentType = "application/json"
+	patPrefix   = "pat_"
+)
```
Comment (arvindh123) [inline auth/api/http/keys/transport.go]: Please revert this changes, it is no more needed

**Review by arvindh123 — CHANGES_REQUESTED** [review]

File: pkg/authn/authn.go [hunk 18]
```diff
@@ -19,4 +41,5 @@ type Session struct {
 //go:generate mockery --name Authentication --output=./mocks --filename authn.go --quiet --note "Copyright (c) Abstract Machines"
 type Authentication interface {
 	Authenticate(ctx context.Context, token string) (Session, error)
+	AuthenticatePAT(ctx context.Context, token string) (Session, error)
```
Comment (arvindh123) [inline pkg/authn/authn.go]: @nyagamunene 
Why do we need separate function `AuthenticatePAT`

File: internal/proto/auth/v1/auth.proto [hunk 19]
```diff
@@ -24,6 +26,11 @@ message AuthNRes {
   string domain_id = 3; // domain id
 }
 
+message AuthNPATRes {
+  string id    = 1; // pat id
+  string user_id = 2; // user id
+}
```
Comment (arvindh123) [inline internal/proto/auth/v1/auth.proto]: I think we can use the same `AuthNRes` with for `AuthNPATRes`.

File: auth/service.go [hunk 20]
```diff
@@ -145,6 +167,21 @@ func (svc service) RetrieveKey(ctx context.Context, token, id string) (Key, erro
 }
 
 func (svc service) Identify(ctx context.Context, token string) (Key, error) {
+	if strings.HasPrefix(token, "pat"+"_") {
```
Comment (arvindh123) [inline auth/service.go]: `Identify` is contains `IdentifyPAT`, so I think we can have only `Authnetication` gRPC enough. 
Please check and give your comments

**Review by arvindh123 — APPROVED** [review]

**Review by arvindh123 — APPROVED** [review]

**Review by dborovcanin — APPROVED** [review]
There is a lot of work that needs to be done, but let's merge the initial implementation and move on. I will open tickets for those tasks today.
